<a href="https://colab.research.google.com/github/pedro-de-bastos/CS166-Modeling-Simulation-and-Decision-Making/blob/main/Pre_Class_Work_CS166_Class_1_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pre-Class Work CS166 Class 1.1


## 1. Arrival and departures rates and exponential distributions
### What are the assumptions of the model?

- The time between consecutive customers joining a queue is exponentially distributed
- The time it takes to serve a customer is normally distributed
- The time it takes for a manager to solve a problem is normally distributed

### What are the variables that represent the state of the system?

- The number of customers in each queue
- The number of servers/queues
- Whether the manager is engaged in solving a particular problem

### What are the rules of the system? (The rules describe how the values of the variables are updated.)

- New customers join lines between 9am and 6pm
- A new customer will join the shortest line (or go directly to a cashier)
- The manager will be engaged in 5% of the cases to solve a problem

### What are the input parameters we need to set before starting the simulation?

- The number of cashiers
- The hyperparameters of our distributions (like the rate $\lambda$ and mean $\mu_{1}$).

### What are the outputs we might want to measure during or at the end of the simulation?

- The average waiting time for each customer
- The percentage of cashiers that were engaged with serving customers at each moment

## How I Applied #Modeling

In this exercise I described the main components of the model, as well as the rules that govern the actors. The model is evidently a simplification of reality, using artifices like the exponential and normal distributions as substitutes to real interactions between agents. The model also simplifies greatly the range of interactions that can occur, and does not allow room for novel and surprising disruptions (like a kid ramming his shopping cart into someone else's cart and causing a ruckus). Nonetheless, this simplification may be good enough to approximate the effect of hiring different numbers of cashiers.

## 2. Simulating exponential random variables

In [ ]:
import numpy as np
import scipy.stats as sts

rate_parameter = 3
distribution = sts.expon(scale=1/rate_parameter)
sample = distribution.rvs(1000)

In [ ]:
print(np.mean(sample))
print(np.var(sample))

0.3222462510122128
0.09517542785324122


The mean of the exponential distribution is $\frac{1}{\lambda}$. Thus, with $\lambda=3$ we expect a mean of $\approx 0.33$ which is what we observe. 

The variance of the exponential distribution is $\frac{1}{\lambda^2}$. Thus we expect a variance of $\frac{1}{9}\approx 0.11$, which is somewhat close to the value of 0.095 we obtained.

## 3. Theoretical Properties of the Exponential Distribution

Firstly, we might deduce that the exponential distribution is the best fit for the data by noting that values around 0 seem to be the most frequent, and that larger values become progressively less frequent, but that the rate of decrease is itself decreasing (we begin with a sharp fall in frequency that then tapers off). This shape is characteristic of the exponential distribution. 

We know that the mean of the exponential distribution is equal to $1/\lambda$. Just by visual inspection of the graph, let us say that if we were to find the average height of the bars we would get something around 3 counts. This would lead us to $3=1/\lambda$, and therefore $\lambda=0\approx0.33$.